### PR TITLE
Dict settings menu: list dictionaries, and allow disabling them

### PR DESCRIPTION
**Pull request by poire-z**
This adds a new dictionary settings menu, with each installed dictionary name, in the order they are used by sdcv, and allows for disabling some of them (this is done by renaming the .ifo file by appending `_Disabled_by_KOReader` to it).
(So I can disable the big greek and latin dictionaries that I don't need for most books: it speeds my fuzzy lookups from 10s to 2s).

### CODE REVIEW

**Comment by Frenzie** [comment]
Note that I've been slacking off on updating sdcv (see [here](https://github.com/koreader/koreader-base/compare/master...Frenzie:fix_sdcv)) but the new version also comes with command line options to that effect (although it seems to be lacking an exclude option?).

https://github.com/Dushistov/sdcv/blob/0cd29823cf2a068417f28ee8ba93761de052b32c/doc/sdcv.1#L34-L35

Also of interest is ordering:
https://github.com/Dushistov/sdcv/blob/0cd29823cf2a068417f28ee8ba93761de052b32c/doc/sdcv.1#L80-L83

And it even keeps a history, although I'm not sure if we should compile it with readline:
https://github.com/Dushistov/sdcv/blob/0cd29823cf2a068417f28ee8ba93761de052b32c/doc/sdcv.1#L75-L79

**Comment by poire-z** [comment]
The `-u` option exists already in our version. But it's somehow easier to do it the way I did, although a bit dirty to rename files.
I don't think we'd need that history/readline things, it's probably easier to do that from koreader itself as we know what we query and we can store it ourselves in a format we like.

**Comment by poire-z** [comment]
Please wait, I'll have a try at using multiple -u, and storing the list of disabled dicts in G_reader_settings.

**Comment by Eduardomb22** [comment]
Can u share those latim and greek dict?

**Comment by poire-z** [comment]
Found them in https://www.mobileread.com/forums/showthread.php?t=256360 (some stardict zips can be downloaded from various posts on this page)

**Comment by poire-z** [comment]
Using multiple -u seems to work fine, and should be cleaner.
But I'll first have to remove the locale_to_utf8() in
https://github.com/koreader/sdcv/blob/4455cf31bb4a424370e29d858dc07cf8d8f5de04/src/sdcv.cpp#L123
otherwise, it ignores (or log "fail to convert to utf8" on my kobo) the french or chinese dictionary names i pass as utf8...

**Comment by poire-z** [comment]
Note that I haven't tested the Android sdcv call change, althought it looks fine if i logger.debug(args), but who knows.

**Comment by Frenzie** [comment]
`table.unpack` changes a table into a return list just like the one that used to be there, so there shouldn't be any issues.

**Comment by poire-z** [comment]
Thanks for the nightlys :)
(My 3 commits could probably have been squashed into a single one)

File: frontend/apps/reader/modules/readerdictionary.lua [hunk 1]
```diff
@@ -40,6 +145,34 @@ function ReaderDictionary:addToMainMenu(menu_items)
...
+                    end
+                    return T(_("Installed dictionaries (%1)"), nb_str)
+                end,
+                sub_item_table = self:genDictionariesMenu(),
+            },
+            {
+                text = _("Info on dictionaries ordering"),
+                callback = function()
+                    UIManager:show(InfoMessage:new{
+                        text = T(_("If you'd like to change the order in which dictionaries are queried (and their results displayed), you can:\n"..
```
Comment (Frenzie) [inline frontend/apps/reader/modules/readerdictionary.lua]: D'oh, I totally overlooked this. See #3151. Currently it needs to be either one long string or use `[[]]` for multiline. I'd like the translation extraction script to do concatenation (which it did where it came from) but in any case I don't have time to look at it today. :-)

File: frontend/apps/reader/modules/readerdictionary.lua [hunk 2]
```diff
@@ -40,6 +145,34 @@ function ReaderDictionary:addToMainMenu(menu_items)
...
+                    if nb_disabled > 0 then
+                        local nb_enabled = nb_available - nb_disabled
+                        nb_str = nb_enabled .. "/" .. nb_available
+                    end
+                    return T(_("Installed dictionaries (%1)"), nb_str)
+                end,
+                sub_item_table = self:genDictionariesMenu(),
+            },
+            {
+                text = _("Info on dictionaries ordering"),
```
Comment (Frenzie) [inline frontend/apps/reader/modules/readerdictionary.lua]: I kind of overlooked this too. I'd say "dictionary order" or "dictionary ordering".

**Comment by Frenzie** [comment]
@poire-z Do you think this could have a result like described here? https://www.mobileread.com/forums/showthread.php?p=3588073#post3588073

> I've downloaded one of the latest nightly from this place: http://build.koreader.rocks/download/nightly/
> 
> Before (with the latest release from 20170821):
> KOReader started immediately. The first time I searched in the dictionary took almost a minute (I have three dictionaries installed).
> 
> After:
> Now it is the opposite: KOReader takes almost a minute to start. First search in dictionary takes no time at all.
> If I delete the dictionaries, KOReader starts as before.

**Comment by poire-z** [comment]
I don't know.
This change will indeed have the dictionary directory browsed on startup, just to read the .ifo files in there.
sdcv would do the same on each call.
It could have that effect if the dict directory on this user's device was "hard" to physicaly read on the sdcard: first time walking the directory hierarchy + reading the ifo could take some time - but once it's done, it's cached in RAM and the following attempts at that (by sdcv or by a restart of koreader) would have a reason to be fast.
The user could check if it's still the same after a koreader restart (without a reboot).
(Or there could be a really bad and huge .ifo, but that would mean not only the first search in dictionary with old version would take a minute, they would all do).
I can't think of anything else.

**Comment by Frenzie** [comment]
@poire-z 

https://www.mobileread.com/forums/showpost.php?p=3612069&postcount=2875
> For people with slow start-up/slow first dictionary access: I had the same problem. It seems it was due to a dictionary containing thousands of small images. I simply deleted the images from the dictionary and now everything is fine (I don’t know what would happen if I tried to access an entry which uses one of those images, I didn’t encounter that situation yet).

**Comment by poire-z** [comment]
I thought this code needed to recurse all directories, as that's what sdcv does anyway, and we don't know how the user would setup dict dirs in `koreader/data/dict/` (he could have make subdirs by language, etc... and sdcv would work with that).
https://github.com/koreader/koreader/blob/2538189473241ecab9013ffcd48fb469fdf20561/frontend/apps/reader/modules/readerdictionary.lua#L23-L51

But may be we could: first look for files matching .ifo, and only if not found, look for directories and recurse in them ? That would avoid walking down subdirectories with images (I assume they are in a subdirectory, aren't they ?). But I guess sdcv would still browse them (and make the OS read the filesystem parts, that's what I guess is slow).

**Comment by Frenzie** [comment]
I'm not aware of any dics with images inside but sounds good to me. :-)

**Comment by poire-z** [comment]
(I think PetitRobert2007 has a dir with a few images and mp3.)

**Comment by poire-z** [comment]
@arooni regarding https://www.mobileread.com/forums/showthread.php?t=328920
see https://github.com/koreader/koreader/pull/3144#issuecomment-344722499 above.
May be one of your (disabled) dicts has many files/subdirs, and sdcv would still walk all that to look for available directories (before not using the ones KOReader tells it to ignore).

**Comment by arooni** [comment]
also see #6064

**Pull request merged**